###################################################################
#####Identifying and removing contaminant ASVs normalized data#####
###################################################################

setwd("~/Documents/Bioinformatics_scripts/R_scripts/Spis_Pver/")

#Fixing file names and IDs
asv = read.table("Input_files/Spis_Pver_ASV_table.txt", header = T,sep = " ")
fileID=read.table("Input_files/Spis_Pver.files", header = F, sep = "\t", row.names = 1)
fileID$ID=gsub("\\_.*", "", fileID$V2)
fileID$ID=gsub("-", "_", fileID$ID)
colnames(asv)=rownames(fileID)[match(colnames(asv), fileID$ID)]
colnames(asv)[767:775]=c("sum","Kingdom","Phylum","Class","Order" ,"Family" ,"Genus","Species","Sequence")

#read in asv table generated by dada2
meta1=read.table("Input_files/SpisPver_metadata.txt", header = T, sep = "\t", row.names = 1)
names(asv)
tax=asv[,c(768:775)]

#Remove samples with low read number
asv.n=apply(asv[,c(1:766)], 2, as.numeric)
asv.o=asv.n[, colSums(asv.n) > 1000] #762 samples with > 1000 reads were retained out of 766 total samples: SMAQ_R1_17 SMAQ_R1_18 SKAU_R2_18 SKAU_R3_2
dim(asv.o)
rownames(asv.o)=rownames(asv)
message(ncol(asv.o)," samples with > 1000 reads were retained out of ", ncol(asv.n), " total samples")
message(" samples descarted: ", colnames(asv.n[, colSums(asv.n) < 1000]))

#Identify and removing contaminant ASVs raw data
asv.r=as.data.frame(sweep(asv.o,2,colSums(asv.o),`/`))
asv.r=transform(asv.r,  Sum = rowSums(asv.r[,2:ncol(asv.r)]))
names(asv.r)
asv.r=transform(asv.r,  SumNegs = rowSums(asv.r[,761:762])) # define negative controls here
names(asv.r)
asv.r=transform(asv.r,  contaFactor=(asv.r$SumNegs/asv.r$Sum)*100)
rownames(asv.r)=rownames(asv)
Conta=subset(asv.r, asv.r$contaFactor > 10)
Conta$Family=asv$Family[match(rownames(Conta), rownames(asv))]
message("Number of total ASVs: ", nrow(asv))
message("Number of identified contaminant ASVs removed from the analysis: ", length(rownames(Conta)), "\n", Conta$Family[1],"\n", Conta$Family[2],"\n", Conta$Family[3],"\n", Conta$Family[4],"\n", Conta$Family[5])

# Export ASV tables wit no outliers, negatives and fixed IDs and only with samples used for the pop genomics analysis
#master=read.tableread.table("Input_files/new_barplots_set.txt")
#master$INDIVIDUALS=gsub("-", "_", master$INDIVIDUALS)
#length(unique(master$INDIVIDUALS)) #644

list=read.table("Input_files/new_barplots_set.txt")
list$V1=gsub("-", "_", list$V1)
length(intersect(list$V1, colnames(asv.o)))

asv.noConta=as.data.frame(subset(asv.o, !rownames(asv.o) %in% rownames(Conta))[,1:760])
asv.popGen=asv.noConta[,which(colnames(asv.noConta) %in% list$V1)]

message("Number of ASVs used in the analysis: ", length(rownames(asv.popGen.f)))
message("Number of samples used in the analysis: ", length(colnames(asv.popGen))) # 660
message("Number of Pver samples used in the analysis: ", length(asv.popGen[, grepl("^P",colnames(asv.popGen))])) # 296
message("Number of Spis samples used in the analysis: ", length(asv.popGen[, grepl("^S",colnames(asv.popGen))])) # 364
asv.popGen.f=merge(asv.popGen, tax, by="row.names", all.x = T)

write.table(asv.popGen.f, "./Input_files/SpisPver_ASVs_noContanoOut.txt",  quote = FALSE, row.names=F, sep = "\t") #define sample range


#subset(master, !INDIVIDUALS  %in% colnames(asv.popGen.f))

# Fix ASV stats table
stats = read.table("Input_files/Spis_Pver_ASV_stats.txt", header = T,sep = "\t")
rownames(stats)=gsub("-", "_", rownames(stats))
rownames(stats)=rownames(fileID)[match(rownames(stats), fileID$ID)]
stats_selection=subset(stats, rownames(stats) %in% colnames(asv.popGen.f))
write.table(stats_selection,"outputs/Spis_Pver_ASV_stats_filtered.txt", quote = F, row.names = T, sep = "\t")

